src/item.cpp

```diff
@@ -6393,186 +6388,161 @@ void item::update_temp( const int temp, const float insulation )
-                item_counter -= loop_diff;
-                // if item is cold and counter is decreasing, we can only be in a
-                // warmer temp, so we're done if it's less than zero, if it's
-                // greater than 600 that means we're still cooling off and need to
-                // transition to frozen.
-                if( item_counter <= 0 ) {
-                    item_tags.erase( "COLD" );
-                    item_counter = 0;
-                    return;
-                }
-            }
-        } else if( is_frozen ) {
-            if( temp <= freeze_point ) {
-                item_counter += loop_diff;
-                if( item_counter > 500 && type->comestible->parasites > 0 ) {
-                    item_tags.insert( "NO_PARASITES" );
-                }
-
-                // counter cannot exceed 600 as frozen, if it does we know
-                // we're done because item is cooling down and cannot cool any
-                // further
-                if( item_counter > 600 ) {
-                    item_counter = 600;
-                    return;
-                }
-            } else {
-                item_counter -= loop_diff;
-                // item is defrosting
-                if( item_counter < 0 ) {
-                    apply_freezerburn(); // removes frozen tag and applies mushy/rot if needed
-                    item_tags.insert( "COLD" );
-                    item_counter = 600;
-                    is_frozen = false;
-                    is_cold = true;
-                }
-            }
-        } else if( temp <= temperatures::cold ) {
-            // see if we need to add a cold tag
-            item_counter += loop_diff;
-            if( item_counter > 600 ) {
-                item_tags.insert( "COLD" );
-                item_counter = 1;
-                is_cold = true;
-            }
-        } else if( temp >= temperatures::hot ) {
-            // if we have item_counters without a tag we were starting to
-            // increment toward cold, remove counters until we don't have any,
-            // then add warm tag and start to increment with WARM tag
-            item_counter -= loop_diff;
-            if( item_counter < 0 ) {
-                item_tags.insert( "WARM" );
-                item_counter = std::max( 600, -item_counter );
-                is_warm = true;
-            }
+    const float env_temperature = ( temp - 32 ) * 0.556 + 273.15;
+    const float old_temperature = 0.1 * temperature;
+    
+    // If no temperature difference then no need to do math.
+    // In practice this happens when the real difference is smaller than 0.05 K due to rounding when the item temperature is saved as integer.
+    // This is not a problem and it is good to stop doing pointless calculations at some point.
+    if( env_temperature == old_temperature ) {
+        return;
+    }
+    const float mass = to_gram( weight() ); // g
+    float true_energy = mass * thermal_energy * 0.001;
```

Comment:
It is not clear (at least for  me) why to use two different variables for the energy, specially if the overflow is not a problem any more. You can probably use only one of them in the entire calculations.

Reply:
Yes the calculations could be done with energy density almost completely (only the energy change part requires actual thermal energy).

The "thermal_energy" should probably be renamed to "thermal_energy_density" and "true_energy" to "thermal_energy" to be more clear.